File: EndGate/EndGate.Core.Client.JS/Assets/Vectors/Vector2d.ts
/// <reference path="../../Scripts/_references.d.ts" />

module EndGate.Core.Assets {
    export class Vector2d implements ITyped {
        public _type: string = "Vector2d";

        public X: number;
        public Y: number;

        constructor(x?: number, y?: number) {
            this.X = x || 0;
            this.Y = y || 0;
        }

        public static Zero(): Vector2d {
            return new Vector2d(0, 0);
        }

        public static One(): Vector2d {
            return new Vector2d(1, 1);
        }

        public ProjectOnto(v: Vector2d): Vector2d {
            return v.Multiply(this.Dot(v) / v.Dot(v));
        }

        public RotateAround(point: Vector2d, angle: number, precision: number = 2) {
            var ca = Math.cos(-angle);
            var sa = Math.sin(-angle);

            return new Vector2d(
                Math.roundTo(ca * (this.X - point.X) - sa * (this.Y - point.Y) + point.X, precision),
                Math.roundTo(sa * (this.X - point.X) + ca * (this.Y - point.Y) + point.Y, precision)
            );
        }

        public Apply(action: (val: number) => number): void {
            this.X = action(this.X);
            this.Y = action(this.Y);
        }

        public Trigger(action: (val: number) => void): void {
            action(this.X);
            action(this.Y);
        }

        public Normalized(): Vector2d {
            var magnitude = this.Magnitude();
            return new Vector2d(this.X / magnitude, this.Y / magnitude);
        }

        public Magnitude(): number {
            return Math.sqrt(this.X * this.X + this.Y * this.Y);
        }

        public Length(): number {
            return this.Magnitude();
        }

        public Dot(v1: Vector2d): number {
            return v1.X * this.X + v1.Y * this.Y;
        }

        public Abs(): Vector2d {
            return new Vector2d(Math.abs(this.X), Math.abs(this.Y));
        }

        public Sign(): Vector2d {
            return new Vector2d(this.X / Math.abs(this.X), this.Y / Math.abs(this.Y));
        }

        public Distance(v1: Vector2d): Vector2d {
            return new Vector2d(Math.abs(v1.X - this.X), Math.abs(v1.Y - this.Y));
        }

        public Add(val: Vector2d): Vector2d;
        public Add(val: number): Vector2d;
        public Add(val: any): Vector2d{
            if (val._type === "Vector2d") {
                return new Vector2d(this.X + val.X, this.Y + val.Y);
            }
            else {
                return new Vector2d(this.X + val, this.Y + val);
            }
        }

        public Multiply(val: Vector2d): Vector2d;
        public Multiply(val: number): Vector2d;
        public Multiply(val: any): Vector2d {
            if (val._type === "Vector2d") {
                return new Vector2d(this.X * val.X, this.Y * val.Y);
            }
            else {
                return new Vector2d(this.X * val, this.Y * val);
            }
        }

        public Subtract(val: Vector2d): Vector2d;
        public Subtract(val: number): Vector2d;
        public Subtract(val: any): Vector2d {
            if (val._type === "Vector2d") {
                return new Vector2d(this.X - val.X, this.Y - val.Y);
            }
            else {
                return new Vector2d(this.X - val, this.Y - val);
            }
        }

        public SubtractFrom(val: Vector2d): Vector2d;
        public SubtractFrom(val: number): Vector2d;
        public SubtractFrom(val: any): Vector2d {
            if (val._type === "Vector2d") {
                return new Vector2d(val.X - this.X, val.Y - this.Y);
            }
            else {
                return new Vector2d(val - this.X, val - this.Y);
            }
        }

        public Divide(val: Vector2d): Vector2d;
        public Divide(val: number): Vector2d;
        public Divide(val: any): Vector2d {
            if (val._type === "Vector2d") {
                return new Vector2d(this.X / val.X, this.Y / val.Y);
            }
            else {
                return new Vector2d(this.X / val, this.Y / val);
            }
        }

        public DivideFrom(val: Vector2d): Vector2d;
        public DivideFrom(val: number): Vector2d;
        public DivideFrom(val: any): Vector2d {
            if (val._type === "Vector2d") {
                return new Vector2d(val.X / this.X, val.Y / this.Y);
            }
            else {
                return new Vector2d(val / this.X, val / this.Y);
            }
        }

        public Negate(): Vector2d {
            return new Vector2d(this.X * -1, this.Y * -1);
        }

        public Equivalent(v: Vector2d): bool {
            return this.X === v.X && this.Y === v.Y;
        }

        public Clone(): Vector2d {
            return new Vector2d(this.X, this.Y);
        }

        // toString override 
        public toString(): string {
            return "(" + this.X + ", " + this.Y + ")";
        }
    }
}

File: EndGate/EndGate.Core.Net/Assets/Vectors/Vector2d.cs
using System;

namespace EndGate.Core.Net.Assets
{
    public class Vector2d
    {
        public Vector2d()
            : this(0, 0)
        {
        }

        public Vector2d(double x, double y)
        {
            X = x;
            Y = y;
        }

        public double X { get; set; }
        public double Y { get; set; }

        public Vector2d ProjectOnto(Vector2d v)
        {
            return (this.Dot(v) / v.Dot(v)) * v;
        }

        public Vector2d RotateAround(Vector2d point, double angle, int precision = 2)
        {
            var ca = Math.Cos(angle);
            var sa = Math.Sin(angle);

            return new Vector2d(
                Math.Round(ca * (X - point.X) - sa * (Y - point.Y) + point.X, precision),
                Math.Round(sa * (X - point.X) + ca * (Y - point.Y) + point.Y, precision)
            );
        }

        public void Apply(Func<double, double> action)
        {
            X = action(X);
            Y = action(Y);
        }

        public void Trigger(Action<double> action)
        {
            action(X);
            action(Y);
        }

        public bool Equivalent(Vector2d v1)
        {
            return v1.X == X && v1.Y == Y;
        }

        public bool IsZero() {
            return this.X == 0 && this.Y == 0;
        }

        /// <summary>
        /// Clones the vector2
        /// </summary>
        /// <returns>A new Vector2s</returns>
        public Vector2d Clone()
        {
            return new Vector2d(X, Y);
        }

        /// <summary>
        /// Creates a Normalized copy of the current Vector2
        /// </summary>
        /// <returns>A new Vector2</returns>
        public Vector2d Normalized()
        {
            var magnitude = Magnitude();
            return new Vector2d(X / magnitude, Y / magnitude);
        }

        /// <summary>
        /// Calculates the magnitude of the Vector2 (X*X + Y*Y)^.5
        /// </summary>
        /// <returns>A new Vector2</returns>
        public double Magnitude()
        {
            return Math.Sqrt(X * X + Y * Y);
        }

        /// <summary>
        /// Equivalent to Magnitude()
        /// </summary>
        /// <returns>A new Vector2</returns>
        public double Length()
        {
            return Magnitude();
        }

        /// <summary>
        /// Calculates the Dot product of the current vector and the one provided
        /// </summary>
        /// <param name="v1">Vector to dot product with</param>
        /// <returns>Dot product</returns>
        public double Dot(Vector2d v1)
        {
            return v1.X * X + v1.Y * Y;
        }

        /// <summary>
        /// Returns a new vector2 which has Math.Abs(X) and Math.Abs(Y) values
        /// </summary>
        /// <returns>A new Vector2</returns>
        public Vector2d Abs()
        {
            return new Vector2d(Math.Abs(X), Math.Abs(Y));
        }

        /// <summary>
        /// Calculates the Sign of the vector (X/|X|, Y/|Y|)
        /// </summary>
        /// <returns>A new Vector2</returns>
        public Vector2d Sign()
        {
            return new Vector2d(X / Math.Abs(X), Y / Math.Abs(Y));
        }

        public Vector2d Distance(Vector2d v1)
        {
            return new Vector2d(Math.Abs(v1.X - X), Math.Abs(v1.Y - Y));
        }

        public Vector2d Add(Number val)
        {
            return this + val;
        }

        public Vector2d Add(Vector2d v1)
        {
            return this + v1;
        }

        public Vector2d Subtract(Number val)
        {
            return this - val;
        }

        public Vector2d SubtractFrom(Number val)
        {
            return val - this;
        }

        public Vector2d Subtract(Vector2d v1)
        {
            return this - v1;
        }

        public Vector2d Multiply(Vector2d v1)
        {
            return this * v1;
        }

        public Vector2d Multiply(Number val)
        {
            return this * val;
        }

        public Vector2d Divide(Number val)
        {
            return this / val;
        }

        public Vector2d DivideFrom(Number val)
        {
            return val / this;
        }

        public Vector2d Divide(Vector2d v1)
        {
            return this / v1;
        }

        public Vector2d Negate()
        {
            return this * -1;
        }

        public static Vector2d operator +(Vector2d v1, Vector2d v2)
        {
            return new Vector2d(v1.X + v2.X, v1.Y + v2.Y);
        }

        public static Vector2d operator +(Vector2d v1, Number val)
        {
            return new Vector2d(v1.X + val, v1.Y + val);
        }

        public static Vector2d operator +(Number val, Vector2d v1)
        {
            return v1 + val;
        }

        public static Vector2d operator -(Vector2d v1, Vector2d v2)
        {
            return new Vector2d(v1.X - v2.X, v1.Y - v2.Y);
        }

        public static Vector2d operator -(Vector2d v1, Number val)
        {
            return new Vector2d(v1.X - val, v1.Y - val);
        }

        public static Vector2d operator -(Number val, Vector2d v1)
        {
            return new Vector2d(val - v1.X, val - v1.Y);
        }

        public static Vector2d operator *(Vector2d v1, Vector2d v2)
        {
            return new Vector2d(v1.X * v2.X, v1.Y * v2.Y);
        }

        public static Vector2d operator *(Vector2d v1, Number val)
        {
            return new Vector2d(v1.X * val, v1.Y * val);
        }

        public static Vector2d operator *(Number val, Vector2d v1)
        {
            return v1 * val;
        }

        public static Vector2d operator /(Vector2d v1, Vector2d v2)
        {
            return new Vector2d(v1.X / v2.X, v1.Y / v2.Y);
        }

        public static Vector2d operator /(Vector2d v1, Number val)
        {
            return new Vector2d(v1.X / val, v1.Y / val);
        }

        public static Vector2d operator /(Number val, Vector2d v1)
        {
            return new Vector2d(val / v1.X, val / v1.Y);
        }

        public static Vector2d operator -(Vector2d v1)
        {
            return v1 * -1;
        }

        public static Vector2d operator ++(Vector2d v1)
        {
            return v1 + 1;
        }

        public static Vector2d operator --(Vector2d v1)
        {
            return v1 - 1;
        }

        public static Vector2d Zero
        {
            get
            {
                return new Vector2d(0, 0);
            }
        }

        public static Vector2d One
        {
            get
            {
                return new Vector2d(1,1);
            }
        }

        public override string ToString()
        {
            return "(" + X + ", " + Y + ")";
        }
    }
}
